src/EFCore.SqlServer/Query/Internal/Translators/SqlServerStringMethodTranslator.cs

```diff
@@ -287,6 +298,37 @@ public SqlServerStringMethodTranslator(ISqlExpressionFactory sqlExpressionFactor
                 method.ReturnType);
         }
 
+        if(PatIndexMethodInfo.Equals(method)
+            || PatIndexMethodInfoWithCollation.Equals(method))
+        {
+            var pattern = arguments[1];
+
+            if (!(pattern is SqlParameterExpression || pattern is SqlConstantExpression))
+            {
+                throw new InvalidOperationException(SqlServerStrings.InvalidPatternForPatIndex);
+            }
+
+            var propertyReference = arguments[2];
+
+            if (propertyReference is not ColumnExpression)
+            {
+                throw new InvalidOperationException(SqlServerStrings.InvalidColumnNameForPatIndex);
+            }
+           
+            return _sqlExpressionFactory.Function(
+                "PATINDEX",
+                new List<SqlExpression>()
+                {
+                    _sqlExpressionFactory.ApplyDefaultTypeMapping(pattern),
+                    arguments.Count == 4
+                        ? new CollateExpression(propertyReference, (string)((SqlConstantExpression)arguments[3]).Value!)
+                        : propertyReference
+                },
+                nullable: true,
+                argumentsPropagateNullability: new[] { false, false },                
```

Comment:
according to the [documentation](https://learn.microsoft.com/en-us/sql/t-sql/functions/patindex-transact-sql)
> If either pattern or expression is NULL, PATINDEX returns NULL.
apparently that is the only way PATINDEX can return null (if the pattern is not found, the result is 0)

you should be able to take advantage of nullability propagation here by setting:
```suggestion
                argumentsPropagateNullability: new[] { true, true },
```

I believe this should result in a (marginally) better query in your `PatIndex_null_pattern`

Reply:
Yes, PATINDEX is [clearly documented](https://learn.microsoft.com/en-us/sql/t-sql/functions/patindex-transact-sql?view=sql-server-ver16) as propagating nulls for both its parameters:

```
If either pattern or expression is NULL, PATINDEX returns NULL.
```

So the `false` here are incorrect, and should be `true` instead.

Reply:
Fixed